File: JavaScript/ExampleTests/common/ValidValuesString.js
import * as tokens from "./tokens.js";

export class ValidValuesString {
  static DNC_STRING = "?DNC?";

  constructor(value = "", isValid = "", notes = "") {
    this.value = value;
    this.isValid = isValid;
    this.notes = notes;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new ValidValuesString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 3, "ValidValues");
    return new ValidValuesString(
      parts[0],
      parts[1],
      parts[2]
    );
  }

  toString() {
    return tokens.token(this.value) + " " + tokens.token(this.isValid) + " " + tokens.token(this.notes);
  }

  equals(other) {
    if (!(other instanceof ValidValuesString)) return false;
    return (this.value === ValidValuesString.DNC_STRING || other.value === ValidValuesString.DNC_STRING || this.value === other.value)
      && (this.isValid === ValidValuesString.DNC_STRING || other.isValid === ValidValuesString.DNC_STRING || this.isValid === other.isValid)
      && (this.notes === ValidValuesString.DNC_STRING || other.notes === ValidValuesString.DNC_STRING || this.notes === other.notes);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: Python/ExampleTests/common/valid_values_string.py
class ValidValuesString:
    def __init__(self, value: str = '', is_valid: str = '', notes: str = ''):
        self.value = value
        self.is_valid = is_valid
        self.notes = notes

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else ''
        )

    def __str__(self):
        return (f'Value={self.value}' + ', ' +
                f'IsValid={self.is_valid}' + ', ' +
                f'Notes={self.notes}')
